### PR TITLE
fix(url): Tree-shake emitting files

### PR DESCRIPTION
**Pull request by artemjackson**
## Rollup Plugin Name: `url`

This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [x] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

### Description

Currently @rollup/plugin-url does not respect Rollup's Tree-shaking.
So all imported files are copied to destination folder even if a parent module has been actually tree-shaked by Rollup.

Simplified example:
```js
// foo.js
import png from "./image.png"

export const foo = 42
export const bar = png

// index.js
import { foo } from "./foo.js"

console.log(foo)
```

This is bundled to:
```
dist/
   index.js
   image.png
```

Despite of `foo.js` was actually tree-shaked:
```js
// index.js
const foo = 42;

console.log(foo);
```

I'm not sure what the PR is a feature or a bugfix, so some help needed to classify it

### CODE REVIEW

**Comment by shellscape** [comment]
Thanks for opening this PR. I'd definitely call this a bugfix. We'll definitely need some tests around this to verify it's working correctly and to prevent regressions, before we can merge.

**Comment by artemjackson** [comment]
I've just noticed that sometimes people just need to import some stuff without actually referencing it in the code, [example](https://rollupjs.org/repl/?version=2.12.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU1QyUyMi4lMkZzdHlsZS5jc3MlNUMlMjIlNUNuJTVDbmNvbnN0JTIwcCUyMCUzRCUyMGRvY3VtZW50LmNyZWF0ZUVsZW1lbnQoJ3AnKSU1Q25wLmNsYXNzTGlzdC5hZGQoJ3NvbWUtY2xhc3MtZnJvbS1zdHlsZXMuY3NzJyklMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJzdHlsZS5jc3MlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwZGVmYXVsdCUyMCU1QyUyMnBhdGgtdG8tY3NzLWZpbGUtZ2VuZXJhdGVkLWJ5LXBsdWdpbiU1QyUyMiUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE).

And this PR  misses such files cuz it relies on `Asset.modules` which are populated only by actually used (referenced in code) modules.

It seems to me that not `file.id` should be checked in `Asset.modules` but file `importer.id`.
In other words assets should be included only if parent module is included in Rollup bundle.

So **DO NOT MERGE** this version :)

I'm going to fix it and provide tests

**Comment by artemjackson** [comment]
Well, I have added test and fixed prettier issues.

Regarding of [this comment](https://github.com/rollup/plugins/pull/435#issuecomment-638302036): I didn't find a way to "teach" Rollup that such import is intentional and should not be tree-shaked. Btw, I had an idea to check all of the imported file parent modules for not being tree-shaked (i.e. to belong to `bundle` `Asset.modules`) but it seems like a bundler job (e.g. to mark modules as tree-shaked  somewhere in [this.getModuleInfo](https://rollupjs.org/guide/en/#thisgetmoduleinfomoduleid-string--moduleinfo)).

So @shellscape do you have any vision on this?

**Comment by lukastaegert** [comment]
rollup/rollup#3663 should give you the solution you are looking for.

**Comment by shellscape** [comment]
@artemjackson please take a look at the commit lukas referenced and let us know if that resolves the issues you mentioned. It looks like this is ready to go now.

**Comment by shellscape** [comment]
_Heads up: Once this hits 60 days without activity the PR will be closed. Please take a look before we reach that point!_
